### ISSUE TITLE
test88 fails with HASH_BER and HASH_SFH

### ISSUE THREAD

**Issue by susnux** [troydhanson/uthash#113]
I am using the 2.0.2 release, my system: openSUSE Leap 42.2 (64bit) gcc 4.8.5.

When running the tests test88 fails for `HASH_FUNCTION=HASH_BER` and `HASH_FUNCTION=HASH_SFH`.

```
uthash/tests% ./all_funcs
[...]
perl ./do_tests
test88 failed
90 tests conducted, 1 failed.
Makefile:111: recipe for target 'run_tests' failed
make: *** [run_tests] Error 1
```

I have attached the `.out` files.

[test88 log for HASH_SFH](https://github.com/troydhanson/uthash/files/883767/test88.out.HASH_FUNCTION.HASH_SFH.txt)
[test88 log for HASH_BER](https://github.com/troydhanson/uthash/files/883773/test88.out.HASH_FUNCTION.HASH_BER.txt)

**Comment by Quuxplusone** [comment]
This is expected because the output of that test (namely the number of `memcmp`s) depends on the hash function (namely the number of hash collisions). The test is only supposed to work with the default `HASH_JEN`; I'm actually shocked that it *does* work with `HASH_SAX` and `HASH_FNV`!

I agree that it would be nice to get the tests green with `make EXTRA_CFLAGS=-DHASH_FUNCTION=HASH_BER -j8`. I'm not sure what's the best approach to get there, though. Any suggestions?

**Comment by susnux** [comment]
> The test is only supposed to work with the default HASH_JEN;

In this case I would exclude test88 from `PROGS` in the Makefile and add something like this:
```
ifeq ($(HASH_FUNCTION),"HASH_JEN")
PROGS += test88
endif
```

Or the other option would be to fix the other hashes or the test to work with the failing hashes.
But as it is intended to only work with JEN I would use the first option.

**Comment by Quuxplusone** [comment]
> `ifeq ($(HASH_FUNCTION),"HASH_JEN")`

@susnux, can you post the command line(s) that you're using to run the tests (e.g. `make -j8` or whatever)?  As implied by my previous message, my default assumption is that we're trying to get these command lines to run green:

    make -j8
    make -j8 EXTRA_CFLAGS=-DHASH_FUNCTION=HASH_BER

(But neither of these command lines interacts with the `HASH_FUNCTION` variable in the Makefile at all. Actually I suspect that the Makefile is a bit borked in that area. I assume you copied the existing codepath dealing with `HASH_MUR`; but I don't understand how *that* codepath is meant to be activated, either.)

**Comment by susnux** [comment]
The whoule process:
```
tar -xf uthash-2.0.2.tar.gz
cd uthash-2.0.2/libut
make -j4
cd tests
make -j4
cd ../../tests
echo -e "y\ny\ny\ny\ny\ny\n" | ./all_funcs
```

**Comment by Quuxplusone** [comment]
Aha, thanks.
Am I correct that the line from `./all_funcs`

    make clean tests_only EXTRA_CFLAGS='-DHASH_FUNCTION=HASH_MUR'

doesn't work either? On my box I get

```
cc  -I../src -g -Wall  -DHASH_FUNCTION=HASH_MUR  -o test1 test1.c
test1.c:24:9: warning: implicit declaration of function 'HASH_MUR' is invalid in C99 [-Wimplicit-function-declaration]
        HASH_ADD_INT(users,id,user);
        ^
[...]
Undefined symbols for architecture x86_64:
  "_HASH_MUR", referenced from:
      _main in test1-be34bc.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make: *** [test1] Error 1
```

and that's been the output ever since the initial commit in 07bca4ce0a21516334e8e82f5792cd6d74adb4ff.

I'm inclined to
- look for a clean way for test88 to `#define HASH_FUNCTION HASH_JEN` at the top of the file, so that it uses that hash function even when the user is trying to `-DHASH_FUNCTION=` something different on the command line
- rip out the `HASH_FUNCTION` variable from the Makefile
- figure out whether `./all_funcs` should be salvaged or ripped out

I don't really understand the issue with `HASH_MUR` and strict aliasing; that's from before my time.

**Comment by Quuxplusone** [comment]
Fixed by fd7365f60261ca9ab7cfbf425e6ae75508c38f06 — thanks!
